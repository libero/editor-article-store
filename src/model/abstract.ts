import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import jsdom from "jsdom";

import { makeSchemaFromConfig } from './utils';
import * as abstractConfig from './config/abstract.config';

export function createAbstractState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = new jsdom.JSDOM('').window.document;

  if (content) {
    xmlContentDocument.body.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema
  });
}

export function createImpactStatementState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = new jsdom.JSDOM('').window.document;

  if (content) {
    xmlContentDocument.body.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema
  });
}
