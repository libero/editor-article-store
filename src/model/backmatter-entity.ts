import { v4 as uuidv4 } from 'uuid';
import {JSONObject} from "./types";

export abstract class BackmatterEntity {
  constructor() {
    this._id = uuidv4();
  }

  get id(): string {
    return this._id;
  }

  protected _id: string;

  protected abstract fromXML(xmlNode: Element): void;
  protected abstract fromJSON(json: JSONObject): void;
  protected abstract createBlank(): void;

  protected createEntity(data?: Element | JSONObject): void {
    if (!data) {
      this.createBlank();
<<<<<<< HEAD
    } else if (data instanceof Element) {
      this.fromXML(data);
    } else {
      this.fromJSON(data);
    }
  }
}
=======
    } else if (data.ownerDocument) {
      this.fromXML(data as Element);
    } else {
      this.fromJSON(data as JSONObject);
    }
  }
}
>>>>>>> origin/667-parse-related-articles
