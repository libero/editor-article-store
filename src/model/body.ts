import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';

import * as bodyConfig from './config/body.config';
import { makeSchemaFromConfig } from './utils';
import xmldom from 'xmldom';
import {Manuscript} from "./manuscript";
import {serializeManuscriptSection} from "../xml-exporter/manuscript-serializer";
import {clearNode} from "../xml-exporter/xml-utils";

export function createBodyState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(bodyConfig.topNode, bodyConfig.nodes, bodyConfig.marks);
  const xmlContentDocument = new xmldom.DOMImplementation().createDocument('', '', null);
  const body = xmlContentDocument.createElement('body');
  if (content) {
    body.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(body),
    schema
  });
}

export function serializeBodyState(xmlDoc: Document, manuscript: Manuscript) {
  const bodyXml = serializeManuscriptSection(manuscript.body, xmlDoc);
  const bodyEl = xmlDoc.querySelector('body') as Element;

  clearNode(bodyEl);
  bodyEl.appendChild(bodyXml);
}
