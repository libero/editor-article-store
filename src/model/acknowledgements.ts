import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';

import * as acknowledgementsConfig from './config/acknowledgements.config';
import { makeSchemaFromConfig } from './utils';
import xmldom from 'xmldom';
import {clearNode} from "../xml-exporter/xml-utils";
import {serializeManuscriptSection} from "../xml-exporter/manuscript-serializer";
import {Manuscript} from "../model/manuscript";

export function createAcknowledgementsState(content?: Element | null): EditorState {
  const xmlContentDocument = new xmldom.DOMImplementation().createDocument('', '', null);
  const ack = xmlContentDocument.createElement('ack');
  if (content) {
    const ackTitle = content.querySelector('title');
    if (ackTitle) {
      content.removeChild(ackTitle);
    }
  }

  const schema = makeSchemaFromConfig(
    acknowledgementsConfig.topNode,
    acknowledgementsConfig.nodes,
    acknowledgementsConfig.marks
  );

  if (content) {
    ack.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(ack),
    schema
  });
}

export function serializeAcknowledgementState(xmlDoc: Document, manuscript: Manuscript) {
  const acknowledgementsXml = serializeManuscriptSection(manuscript.acknowledgements, xmlDoc);
  let acknowledgementsEl = xmlDoc.querySelector('ack') as Element;

  if (!acknowledgementsEl) {
    acknowledgementsEl = xmlDoc.createElement('ack');
    acknowledgementsEl.setAttribute('id', 'ack');
    const ackTitleEl = xmlDoc.createElement('title');

    ackTitleEl.appendChild(xmlDoc.createTextNode('Acknowledgements'));
    acknowledgementsEl.appendChild(ackTitleEl);
    xmlDoc.querySelector('back')?.appendChild(acknowledgementsEl);
  }

  const ackTitleEl = acknowledgementsEl.firstChild as Element;
  clearNode(acknowledgementsEl);
  acknowledgementsEl.appendChild(ackTitleEl);
  acknowledgementsEl.appendChild(acknowledgementsXml);
}
