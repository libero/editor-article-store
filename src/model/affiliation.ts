import { BackmatterEntity } from "./backmatter-entity";
import { JSONObject } from "./types";

export class Affiliation extends BackmatterEntity {
  label: string | undefined;
  institution: { name: string; } | undefined;
  address: { city: string; } | undefined;
  country: string | undefined;

  constructor(data?: JSONObject | Element) {
    super();
    this.createEntity(data);
  }

  protected createBlank(): void {
    this.label = '';
    this.country = '';
    this.institution = { name: '' };
    this.address = { city: '' };
  }

  protected fromXML(_xml: Element): void {}

  protected fromJSON(json: JSONObject): void {
    this._id = (json._id as string) || this.id;
    this.label = json.label as string;
    this.country = json.country as string;
    this.institution = json.institution as { name: string };
    this.address = json.address as { city: string };
  }
}