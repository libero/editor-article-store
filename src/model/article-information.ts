import {EditorState} from 'prosemirror-state';
import {DOMParser as ProseMirrorDOMParser} from 'prosemirror-model';
import {BackmatterEntity} from "./backmatter-entity";
import * as licenseTextConfig from './config/license-text.config';
import {getTextContentFromPath, makeSchemaFromConfig} from "./utils";
import {JSONObject} from "./types";
import {Person} from "./person";
import {get} from 'lodash';
import xmldom from "xmldom";

export const LICENSE_CC_BY_4 = 'CC-BY-4';
export const LICENSE_CC0 = 'CC0';

export class ArticleInformation extends BackmatterEntity {

  articleType: string = '';
  dtd: string = '';
  articleDOI: string = '';
  publisherId: string = '';
  volume: string = '';
  elocationId: string = '';
  subjects: Array<string> = [];
  publicationDate: string = '';
  licenseType: string = '';
  copyrightStatement: string = '';
  licenseText?: EditorState;

  constructor(data?: JSONObject | Element, authors?: Person[]) {
    super();
    this.createEntity(data);
    if (get(data, 'ownerDocument') && authors) {
      this.updateCopyrightStatementFromXml(data as Element, authors);
    }
  }

  public static createLicenseEditorState(content?: Element | null): EditorState {
    const schema = makeSchemaFromConfig(licenseTextConfig.topNode, licenseTextConfig.nodes, licenseTextConfig.marks);
    const xmlContentDocument = new xmldom.DOMImplementation().createDocument('', '', null);
    if (content) {
      xmlContentDocument.appendChild(content);
    }

    return EditorState.create({
      doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
      schema
    });
  }

  public updateCopyrightStatement(authors: Person[]): void {
    if (this.licenseType === LICENSE_CC_BY_4) {
      const authorNamesSection =
        authors.length === 1
          ? authors[0].lastName
          : authors.length === 2
          ? `${authors[0].lastName} and ${authors[1].lastName}`
          : authors.length > 2
            ? `${authors[0].lastName} et al`
            : '';

      const date = new Date(this.publicationDate);
      this.copyrightStatement = `Â© ${!isNaN(date.getTime()) ? date.getFullYear() : ''}, ${authorNamesSection}`;
    } else {
      this.copyrightStatement = '';
    }
  }

  protected fromJSON(json: JSONObject): void {
    this.articleType = json.articleType as string;
    this.dtd = json.dtd as string;
    this.articleDOI = json.articleDOI as string;
    this.elocationId = json.elocationId as string;
    this.volume = json.volume as string;
    this.publisherId = json.publisherId as string;
    this.subjects = (json.subjects as string[]) || [];

    this.licenseType = json.licenseType as string;
    this.publicationDate = json.publicationDate as string;
    const schema = makeSchemaFromConfig(licenseTextConfig.topNode, licenseTextConfig.nodes, licenseTextConfig.marks);
    this.licenseText = EditorState.fromJSON({ schema }, json.licenseText as JSONObject);
  }

  protected fromXML(xmlNode: Element): void {
    this.articleType = getTextContentFromPath(xmlNode, 'article-meta subj-group[subj-group-type="heading"]');
    const articleEl = xmlNode.ownerDocument.querySelector('article');
    this.dtd = articleEl ? articleEl.getAttribute('dtd-version') || '' : '';
    this.articleDOI = getTextContentFromPath(xmlNode, 'article-meta article-id[pub-id-type="doi"]');
    this.elocationId = getTextContentFromPath(xmlNode, 'article-meta elocation-id');
    this.volume = getTextContentFromPath(xmlNode, 'article-meta volume');
    this.publisherId = getTextContentFromPath(xmlNode, 'article-meta article-id[pub-id-type="publisher-id"]');
    this.subjects = Array.from(xmlNode.querySelectorAll('subj-group[subj-group-type="subject"] subject')).map(
      (el: Element) => el.textContent || ''
    );

    this.licenseType = this.getLicenseTypeFromXml(xmlNode);
    this.publicationDate = this.getPublicationDateFromXml(xmlNode);
    this.licenseText = ArticleInformation.createLicenseEditorState(
      xmlNode.querySelector('article-meta permissions license license-p')
    );
  }

  protected createBlank(): void {
    this.articleType = '';
    this.dtd = '';
    this.articleDOI = '';
    this.elocationId = '';
    this.volume = '';
    this.publisherId = '';
    this.subjects = [];

    this.licenseType = '';
    this.publicationDate = '';
    this.licenseText = ArticleInformation.createLicenseEditorState();
  }

  private getLicenseTypeFromXml(doc: Element): string {
    const licenseEl = doc.querySelector('article-meta permissions license');
    if (!licenseEl) {
      return '';
    }
    const href = licenseEl.getAttribute('xlink:href');
    if (href === 'http://creativecommons.org/licenses/by/4.0/') {
      return LICENSE_CC_BY_4;
    }
    if (href === 'http://creativecommons.org/publicdomain/zero/1.0/') {
      return LICENSE_CC0;
    }

    return '';
  }

  private getPublicationDateFromXml(xmlNode: Element): string {
    let publicationDate = '';
    const pubDateNode = xmlNode.querySelector('pub-date[date-type="pub"][publication-format="electronic"]');
    if (pubDateNode) {
      publicationDate = [
        get(pubDateNode.querySelector('year'), 'textContent', ''),
        get(pubDateNode.querySelector('month'), 'textContent', ''),
        get(pubDateNode.querySelector('day'), 'textContent', ''),
      ].join('-');
    }

    return publicationDate;
  }

  private updateCopyrightStatementFromXml(xmlNode: Element, authors: Person[]): void {
    const copyrightStatementFromXml = getTextContentFromPath(xmlNode, 'article-meta permissions copyright-statement');
    if (copyrightStatementFromXml) {
      this.copyrightStatement = copyrightStatementFromXml;
    } else {
      this.updateCopyrightStatement(authors);
    }
  }
}
