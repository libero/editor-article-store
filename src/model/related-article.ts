import { BackmatterEntity } from './backmatter-entity';
import { JSONObject } from './types';
import { Manuscript } from '../model/manuscript';
import { DOMImplementation } from 'xmldom';

export class RelatedArticle extends BackmatterEntity {
    public articleType: string = '';
    public href: string = '';

    constructor(data?: JSONObject | Element) {
        super();
        this.createEntity(data);
    }

    public clone(): RelatedArticle {
        const newArticle = new RelatedArticle();
        newArticle._id = this._id;
        newArticle.articleType = this.articleType;
        newArticle.href = this.href;
        return newArticle;
    }

    public toXml(listIndex: number): Element {
        const xmlDoc = new DOMImplementation().createDocument(null, null);
        const articleXml = xmlDoc.createElement('related-article');
        articleXml.setAttribute('ext-link-type', 'doi');
        articleXml.setAttribute('id', `ra${listIndex}`);
        articleXml.setAttribute('related-article-type', this.articleType);
        articleXml.setAttribute('xlink:href', this.href);
        return articleXml;
    }

    protected fromXML(xmlNode: Element): void {
        this._id = xmlNode.getAttribute('id') || this.id;
        this.articleType = xmlNode.getAttribute('related-article-type') || '';
        this.href = xmlNode.getAttribute('xlink:href') || '';
    }

    protected fromJSON(json: JSONObject): void {
        this._id = (json._id as string) || this.id;
        this.articleType = json.articleType as string;
        this.href = json.href as string;
    }

    protected createBlank(): void {
        this.articleType = '';
        this.href = '';
    }
}

export function createRelatedArticleState(relatedArticlesXml: Element[]): RelatedArticle[] {
    return relatedArticlesXml.map((xml: Element) => new RelatedArticle(xml));
}

export function serializeRelatedArticles(xmlDoc: Document, manuscript: Manuscript) {
    xmlDoc.querySelectorAll('article-meta > related-article').forEach((el: Element) => el.parentNode!.removeChild(el));

    const articleMeta = xmlDoc.querySelector('article-meta');
    manuscript.relatedArticles.forEach((article: RelatedArticle, index: number) => {
        articleMeta!.appendChild(article.toXml(index + 1));
    });
}
