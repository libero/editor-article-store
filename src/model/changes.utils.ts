import { cloneDeepWith } from 'lodash';
import { EditorState } from 'prosemirror-state';
import {Manuscript} from './manuscript';
import {ProsemirrorChange} from "./history/prosemirror-change";
import { Change } from './history/change';
import { BatchChange } from "./history/batch-change";
import { AddObjectChange } from "./history/add-object-change";
import { DeleteObjectChange } from './history/delete-object-change';
import { BackmatterEntity } from './backmatter-entity';
import { JSONObject } from "./types";
import { RelatedArticle } from './related-article';
import {UpdateObjectChange} from "./history/update-object-change";
import {Person} from "./person";
import { Keyword } from './keyword';
import { Affiliation } from './affiliation';

export function manuscriptEntityToJson<T>(object: T): JSONObject {
  return cloneDeepWith(object, (value) => {
    if (value instanceof EditorState) {
      return value.toJSON();
    }
  });
}

export function deserializeChanges(changesJson: JSONObject[]): Array<Change>  {
  return changesJson.map((changeData) => {
    switch (changeData.type) {
      case 'prosemirror':
        return ProsemirrorChange.fromJSON(changeData);
      case 'batch':
        return BatchChange.fromJSON(changeData);
      case 'add-object':
        return AddObjectChange.fromJSON(changeData);
      case 'delete-object':
        return DeleteObjectChange.fromJSON(changeData);
      case 'update-object':
        return UpdateObjectChange.fromJSON(changeData);
      default:
        return null;
    }
  }).filter(Boolean) as Change[];
}

export function cloneManuscript<T>(manuscript: T): T {
  const cloneCustomizer = (value: any): EditorState | undefined => (value instanceof EditorState ? value : undefined);
  return cloneDeepWith(manuscript, cloneCustomizer);
}

export function deserializeBackmatter(path: string, json: JSONObject): BackmatterEntity {
  if (path.indexOf('affiliations') >= 0) {
    return new Affiliation(json);
  }

  if (path.indexOf('authors') >= 0) {
    return new Person(json);
  }

  // if (path.indexOf('references') >= 0) {
  //   return new Reference(json);
  // }

  if (path.indexOf('relatedArticles') >= 0) {
    return new RelatedArticle(json);
  }

  if (path.match(/keywordGroups\.[^.]+\.keywords/)) {
    return new Keyword(json);
  }
  throw new Error(`deserialization of backmatter entity for ${path} is not implemented or provided path is invalid`);
}

export function applyChangesToManuscript(manuscript: Manuscript, changesJson: JSONObject[]): Manuscript {
  const allChanges = new BatchChange(deserializeChanges(changesJson) as Change[]);
  return allChanges.applyChange(manuscript);
}
