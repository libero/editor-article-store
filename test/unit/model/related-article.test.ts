/**
 * @jest-environment jsdom
 */
import {RelatedArticle} from "../../../src/model/related-article";

jest.mock('uuid', () => ({
  v4: () => 'unique_id'
}));

describe('Related Article model', () => {
  it('creates a related article from object ', () => {
    expect(new RelatedArticle({articleType: 'SOME_ARTICLE_TYPE', href: 'URL'})).toMatchSnapshot();
  });

  it('creates a related article from xml', () => {
    const el = document.createElement('related-article');
    el.setAttribute('related-article-type', 'commentary-article');
    el.setAttribute('xlink:href', '10.7554/eLife.48498');
    expect(new RelatedArticle(el)).toMatchSnapshot();
  });

  it('creates an empty article if attributes are missing', () => {
    const el = document.createElement('related-article');
    expect(new RelatedArticle(el)).toMatchSnapshot();
  });


  it('creates an empty related article', () => {
    expect(new RelatedArticle()).toMatchSnapshot();
  });

});
