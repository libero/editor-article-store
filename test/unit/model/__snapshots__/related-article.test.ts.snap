// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Related Article model creates a related article from object  1`] = `
RelatedArticle {
  "_id": "unique_id",
  "articleType": "SOME_ARTICLE_TYPE",
  "href": "URL",
}
`;

exports[`Related Article model creates a related article from xml 1`] = `
RelatedArticle {
  "_id": "unique_id",
  "articleType": "commentary-article",
  "href": "10.7554/eLife.48498",
}
`;

exports[`Related Article model creates an empty article if attributes are missing 1`] = `
RelatedArticle {
  "_id": "unique_id",
  "articleType": "",
  "href": "",
}
`;

exports[`Related Article model creates an empty related article 1`] = `
RelatedArticle {
  "_id": "unique_id",
  "articleType": "",
  "href": "",
}
`;
